src/api/libcellml/annotator.h

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <any>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "libcellml/exportdefinitions.h"
+#include "libcellml/logger.h"
+#include "libcellml/types.h"
+
+namespace libcellml {
+
+using VariablePair = std::pair<VariablePtr, VariablePtr>;
+using UnitItem = std::pair<UnitsPtr, size_t>;
+using AnyItem = std::pair<std::uint64_t, std::any>;
+using ItemList = std::map<std::string, AnyItem>;
+/**
+ * @brief The Annotator class.
+ *
+ * The Annotator class is for representing a CellML Annotator.
+ */
+class LIBCELLML_EXPORT Annotator: public Logger
+{
+public:
+    enum Type
+    {
+        COMPONENT,
+        COMPONENT_REF,
+        CONNECTION,
+        ENCAPSULATION,
+        IMPORT,
+        ISSUE,
+        MAP_VARIABLES,
+        MODEL,
+        RESET,
+        RESET_VALUE,
+        TEST_VALUE,
+        UNIT,
+        UNITS,
+        VARIABLE,
+    };
+
+    ~Annotator() override; /**< Destructor */
+    Annotator(const Annotator &rhs) = delete; /**< Copy constructor */
+    Annotator(Annotator &&rhs) noexcept = delete; /**< Move constructor */
+    Annotator &operator=(Annotator rhs) = delete; /**< Assignment operator */
+
+    /**
+     * @brief Create a @c Annotator object.
+     *
+     * Factory method to create a @c Annotator.  Create a
+     * annotator with::
+     *
+     *   AnnotatorPtr annotator = libcellml::Annotator::create();
+     *
+     * @return A smart pointer to an @c Annotator object.
+     */
+    static AnnotatorPtr create() noexcept;
+
+    /**
+     * @brief Builds the internal map of ids to items for the model provided.
+     *
+     * @param model A @c ModelPtr model to build map for.
+     */
+    void build(const ModelPtr &model);
```

Comment:
can we leverage the `build` method to add IDs to any objects that don't have them? Seems like that capability should be somewhere in the annotator class rather than (as well as?) the current addition of IDs in the printer.